added nat configuration

d78ui98 opened the pull request:
I have added `NAT` configuration to the dhcp server.

d78ui98 committed:
added nat configuration

dashohoxha commented on dhcp/install.sh:
```diff
@@ -2,12 +2,33 @@
 
-#configuration
+# fetching interface name
+INTERNET=$(ifconfig | grep -m 1 RUNNING | cut -d':' -f1)
```
I understand that you have copied this from some example, and in the case of example it may give a correct result, but not always.
Try this one which is safer: `ip route | grep default | cut -d' ' -f5`

d78ui98 replied:
okay. I will now prefer `ip route` over `ifconfig`.

dashohoxha replied:
> okay. I will now prefer ip route over ifconfig

It is not a matter of preference, your first implementation is not guaranteed to work in all the cases.

dashohoxha commented on dhcp/install.sh:
```diff
@@ -2,12 +2,33 @@
 apt --yes update
 apt --yes --install-recommends install dnsmasq
+export DEBIAN_FRONTEND=noninteractive
+apt -yq --yes install iptables-persistent
```
Here `-yq` is not needed.

d78ui98 replied:
`-yq` flag is required to handle `DEBIAN_FRONTEND` variable.

dashohoxha replied:
> -yq flag is required to handle DEBIAN_FRONTEND variable.

Are you sure about this? Where did you find this information?

d78ui98 replied:
I can share the source - https://serverfault.com/a/227194

dashohoxha replied:
Look up in the man page the meaning of `-yq` and then remove it.
You have spent me 1 our only for this, and it is not normal that I spend in the project as much time as you do.

dashohoxha commented on dhcp/install.sh:
```diff
@@ -2,12 +2,33 @@
-#configuration
+# fetching interface name
+INTERNET=$(ifconfig | grep -m 1 RUNNING | cut -d':' -f1)
+LOCAL=$(ifconfig | grep -m 2 RUNNING | cut -d':' -f1 | tail -n1)
```
Here try something like this:
`ip route | grep -v default | cut -d' ' -f3 | grep -v $INTERNET | head -1`

dashohoxha commented on dhcp/install.sh:
```diff
@@ -2,12 +2,33 @@
+LOCAL=$(ifconfig | grep -m 2 RUNNING | cut -d':' -f1 | tail -n1)
+
+# configuration
 echo "dhcp-range=tftp,${NETWORK}.250,${NETWORK}.254" >> /etc/dnsmasq.conf 
```
Are you sure we need this line with `tftp`? What does the documentation of `dnsmasq` say about it?

d78ui98 replied:
yes, the documentation says If dnsmasq is providing a TFTP service then only the filename is required to enable network booting. 
Some comment on stackoverflow gave an explanation to this line and concluded that it would reduce time required for network boot.

dashohoxha replied:
> yes, the documentation says If dnsmasq is providing a TFTP service then only the filename is required to enable network booting.

We don't need TFTP service on the DHCP server.

> Some comment on stackoverflow gave an explanation to this line and concluded that it would reduce time required for network boot.

You cannot trust all the comments on stackoverflow. Couldn't you find anything on the documentation of dnsmasq?

dashohoxha replied:
> We don't need TFTP service on the DHCP server.

Please remove this line and fix the configuration of dnsmasq.

dashohoxha commented on dhcp/install.sh:
```diff
@@ -2,12 +2,33 @@
+# installation
 apt --yes update
 apt --yes --install-recommends install dnsmasq
+export DEBIAN_FRONTEND=noninteractive
```
What happens if you remove (or comment out) this line?

d78ui98 replied:
This line is required to handle interactive post-install configuration. If we remove this we won't be able to handle that with our script. As a result we won't be able to install `iptables-persistent`

dashohoxha replied:
> This line is required to handle interactive post-install configuration. If we remove this we won't be able to handle that with our script. As a result we won't be able to install iptables-persistent

Have you tried this?

d78ui98 replied:
yes. I can share a screenshot of what happens if I remove that line. The provision script just freezes - 
![sa](https://user-images.githubusercontent.com/27950739/40978631-82047cea-68f1-11e8-8f19-ee51b756a8ad.jpg)

dashohoxha replied:
> yes. I can share a screenshot of what happens if I remove that line.

If I try to remove it and there is no problem, then **you** are going to have problems. OK?

d78ui98 replied:
Sure. I can make it easy for you. I have created a video which has both cases(line commented and line not commented) https://youtu.be/WQHwKu9ebYA

dashohoxha replied:
Ok, do it like this:
```
DEBIAN_FRONTEND=noninteractive apt --yes install iptables-persistent
```

dashohoxha commented on dhcp/install.sh:
```diff
@@ -2,12 +2,33 @@
+iptables-save > /etc/iptables.rules
+
+# restarting service	
+service dnsmasq restart
```
You can move this line after changing the configuration (or place the configuration change above it), just to keep them together.

d78ui98 committed:
ip route inplace of ifconfig

using ip route in place of ifconfig, adding proper comment, clubbing up code together

d78ui98 committed:
removed -yq

d78ui98 committed:
removed tftp services

dhcp server now does not offer tftp services and interactive yes no box is handled by single line

dashohoxha reviewed (approved):
I am not sure that the default configuration of dnsmasq is correct (without you customizing it).
But you can merge this branch if you wish and fix this problem later, if needed.
